pkg/scheduler/framework/runtime/batch.go

```diff
@@ -0,0 +1,226 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package runtime
+
+import (
+	"context"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	fwk "k8s.io/kube-scheduler/framework"
+	"k8s.io/kubernetes/pkg/scheduler/apis/config"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+	"k8s.io/kubernetes/pkg/scheduler/metrics"
+)
+
+type PodSignatureFunc func(p *v1.Pod) string
+
+// Only needed until we connect signatures
+func noBatchSignatures(p *v1.Pod) string { return "" }
+
+// OpportunisticBatching caches results from filtering and scoring when possible to optimize
+// scheduling of common pods.
+type OpportunisticBatch struct {
+	profileName   string
+	state         *batchState
+	lastCycle     schedulingCycle
+	signatureFunc PodSignatureFunc
+	handle        fwk.Handle
+
+	// Used primarily for tests, count the total pods we
+	// have successfully batched.
+	batchedPods int64
+}
+
+type batchState struct {
+	signature    string
+	sortedNodes  framework.SortedScoredNodes
+	creationTime time.Time
+}
+
+type schedulingCycle struct {
+	cycleCount int64
+	chosenNode string
+	succeeded  bool
+}
+
+const (
+	maxBatchAge = 500 * time.Millisecond
+)
+
+// Provide a hint for the pod based on filtering an scoring results of previous cycles. Caching works only for consecutive pods
+// with the same signature that are scheduled in 1-pod-per-node manner (otherwise previous scores could not be reused).
+// It's assured by checking the top rated node is no longer feasible (meaning the previous pod was successfully scheduled and the
+// current one does not fit). Note that this call does not modify state.
```

Comment:
It does (pop the top node)

```suggestion
// current one does not fit). 
```